### ISSUE TITLE
Request was for a topic or partition that does not exist on this broker

### ISSUE THREAD

**Issue by yitian108** [IBM/sarama#2743]
##### Description

1. Set auto.create.topics.enable = true

2. Initialize a consumer with the following configuration
```go
        cfg := sarama.NewConfig()
	cfg.Version = sarama.V2_5_1_0
	cfg.ClientID = clientId
	cfg.Consumer.Return.Errors = false
	cfg.Consumer.Offsets.AutoCommit.Enable = false
	cfg.Consumer.MaxWaitTime = 200 * time.Millisecond
	cfg.Consumer.Group.Rebalance.Strategy = sarama.BalanceStrategySticky
	cfg.Consumer.Group.Session.Timeout = time.Second * 20
	cfg.Consumer.Group.Heartbeat.Interval = time.Second * 6
```

3. Call `Partitions(topic)` method
```go
    newClient, err := sarama.NewClient(brokerList, cfg)
    if err != nil {
		log.ERROR("New Client for consumer group failed: ", err)
		SysLog.Println(err)
		return nil, err
	}

	admin, err := sarama.NewClusterAdminFromClient(newClient)
	if err != nil {
		log.ERROR("New Admin client for Validating Partitions for Group %s error: %s", groupId, err.Error())
		return nil, err
	}

       partitionMutex.Lock()
	defer partitionMutex.Unlock()

       partitions, err := newClient.Partitions(topic)
       if err != nil {
		log.ERROR("Get Partitions for Validating Partition Topic %s error: %s", topic, err.Error())
		return nil, err
	}
```

   Note: _var partitionMutex sync.RWMutex_

4. Occasionally, it reports the exception by calling Partitions():
```
Request was for a topic or partition that does not exist on this broker
```

##### Versions

| Sarama | Kafka | Go  |
| ------ | ----- | --- |
|  1.37  |  3.6   | 1.18 |

Note: the kafka based on the image: bitnami/kafka:3.6

##### Logs

<details><summary>logs: CLICK ME</summary>
<p>

```
panic: [2023-12-12...] kafka server: Request was for a topic or partition that does not exist on this broker

```

</p>
</details>

##### Additional Context

It usually happens while creates consumer client for peer to peer consumption model(serval consumers share one consumer group).
I found the issue has been mentioned [here](https://github.com/IBM/sarama/issues/1361), and it reported the issued has been solved, Does this related to the configuration `cfg.Version = sarama.V2_5_1_0` or the incompatible  broker version 3.6 with sarama version 1.37?

Is there any good way to solve this?

**Comment by yitian108** [comment]
@RazOmessi, could you please have a look this? Thanks

**Comment by yitian108** [comment]
@RazOmessi, could you please have a look at this? Thanks

**Comment by yitian108** [comment]
@dnwe , have you ever encountered this problem? Thanks

**Comment by dnwe** [comment]
@yitian108 with AllowAutoTopicCreation = true in the client's request, and assuming it is correctly enabled broker side, that shouldl cause the broker to automatically create the topic and the initial metadata response you get back will list the topic and partitions, but an initial leaderless error (whilst it's mid creation) which should cause the client to retry the metadatarequest https://github.com/IBM/sarama/blob/6ab4e2b0dda33ebebf1b160b76a8239acef4987d/client.go#L1047-L1051 up to a maximum of cfg.Metadata.Retry.Max times

You can set `sarama.Logger` to a `log.Logger` to capture Sarama debug output whilst running your test and that may more clearly indicate what code path you're hitting

It's not obvious from what you've provided where the panic() is being called, but presumably in your own code?

**Comment by yitian108** [comment]
@dnwe, thanks for your response. Yes, I did set `auto.create.topics.enable = true` on broker side first, should I set `AllowAutoTopicCreation = true` again on client side? I also set the environment variable `auto.leader.rebalance.enable = true` in this version, I am not sure whether it is associated with leaderless error as you mentioned. Anyway, even if I set the two broker environment variable to true, I can set cfg.Metadata.Retry.Max.times to avoid this problem, right?

I can open the sarama.Logger to output a log file, but what I worried about is it will become larger and larger with the client long running, and exhause the disk eventually if I do not set the roll log mechanism.

Last but least, I am also not sure it was the unmatched version cause this issue as I said first "_Does this related to the configuration cfg.Version = sarama.V2_5_1_0 or the incompatible broker version 3.6 with sarama version 1.37?_", I want to update the broker version to 3.6, but if I still use the sarama version 1.37(which version seems to support the highest version is V3.2 of kafka broker), will this update cause other issues?

**Comment by dnwe** [comment]
@yitian108 auto topic creation is included in the MetadataRequest as long as your config.Version is V0_11_0_0 or newer, so V2_5_1_0 will be fine for that. In general I would recommend that if you know the minimum kafka version your app will be connecting to then you should set the Version field to match if possible

cfg.Metadata.Retry.Max defaults to 3 so unless you're changing it, your client should already be making 3 metadata requests for the topic before returning "Request was for a topic or partition that does not exist on this broker"

**Comment by yitian108** [comment]
@dnwe , ok, I will set the cfg.Metadata.Retry.Max to 6 to see if the app will report the issue "Request was for a topic or partition that does not exist on this broker" again. After all, it is occasional.

I have reviewed the release notes beyond version 1.40 of Sarama. I believe that these numerous updates will address the issue mentioned in the title. I tested the application by setting config.Version to V3_6_0_0, but encountered issues when the broker version was 2.5.0. Therefore, I have reverted the Version back to V2_5_1_0 for now.
